### PR TITLE
Add cache on the graph of the home page

### PR DESCRIPTION
**Pull request by gillesbourgeat**
Add cache on the graph of the home page.
Possibility to disable cache or change ttl cache, with the configuration variable admin_cache_home_stats_ttl

Default value for admin_cache_home_stats_ttl = 600

If admin_cache_home_stats_ttl == 0, then the cache is disabled.

Add btn for clear cache in graph toolbar.

Cheers,

### CODE REVIEW

File: core/lib/Thelia/Controller/Admin/HomeController.php [hunk 1]
```diff
@@ -74,61 +86,100 @@ public function loadStatsAjaxAction()
             return $response;
         }
 
-        $data = new \stdClass();
+        $cacheEnable = intval(ConfigQuery::read("admin_cache_home_stats_enable", '1'));
```
Comment (lunika) [inline core/lib/Thelia/Controller/Admin/HomeController.php]: I see you use a lot the config table in your PR and it's a good idea.

It's a good practice to create an alias when you want to retrieve the value of a specific config. Doing this you prevent typo and you can easily change the name of your variable without refactoring all your code. Moreover, each time you cast the result in int. Do this in the alias method.

++

File: core/lib/Thelia/Controller/Admin/HomeController.php [hunk 2]
```diff
@@ -74,61 +86,100 @@ public function loadStatsAjaxAction()
...
+        $cacheContent = false;
 
-        /* sales */
-        $saleSeries = new \stdClass();
-        $saleSeries->color = $this->getRequest()->query->get('sales_color', '#adadad');
-        $saleSeries->data = OrderQuery::getMonthlySaleStats(
-            $this->getRequest()->query->get('month', date('m')),
-            $this->getRequest()->query->get('year', date('Y'))
-        );
+        if ($cacheEnable) {
```
Comment (lunika) [inline core/lib/Thelia/Controller/Admin/HomeController.php]: variable cacheEnable is used only once, so use directly the ConfigQuery method here.

File: core/lib/Thelia/Controller/Admin/HomeController.php [hunk 3]
```diff
@@ -74,61 +86,100 @@ public function loadStatsAjaxAction()
...
 
-        /* new customers */
-        $newCustomerSeries = new \stdClass();
-        $newCustomerSeries->color = $this->getRequest()->query->get('customers_color', '#f39922');
-        $newCustomerSeries->data = CustomerQuery::getMonthlyNewCustomersStats(
-            $this->getRequest()->query->get('month', date('m')),
-            $this->getRequest()->query->get('year', date('Y'))
-        );
+            $cacheKey = self::STATS_CACHE_KEY . $context;
+            $cacheExpire = intval(ConfigQuery::read("admin_cache_home_stats_ttl", '600'));
```
Comment (lunika) [inline core/lib/Thelia/Controller/Admin/HomeController.php]: same here for the alias

**Comment by gillesbourgeat** [comment]
Hello,

is there really a need to create a static method in config class for a variable that will  be used only here ?

**Comment by lunika** [comment]
@gillesbourgeat yes it is. Tomorrow this variable can be used anywhere and if we use a string we can introduce a typo and it's hard to detect this type of error.

File: core/lib/Thelia/Controller/Admin/HomeController.php [hunk 4]
```diff
@@ -38,61 +50,99 @@ public function loadStatsAjaxAction()
...
-            $newCustomerSeries,
-            $orderSeries,
-            $firstOrderSeries,
-            $cancelledOrderSeries,
-        );
-
-        $json = json_encode($data);
-
-        return $this->jsonResponse($json);
+        $cacheExpire = intval(ConfigQuery::getAdminCacheHomeStatsTTL());
```
Comment (lunika) [inline core/lib/Thelia/Controller/Admin/HomeController.php]: with this static method you can control what is returned, please put your intval in the method ```getAdminCacheHomeStatsTTL()```

Comment (lunika) [inline core/lib/Thelia/Controller/Admin/HomeController.php]: Better, you can cast it in boolean if you want

Comment (gillesbourgeat) [inline core/lib/Thelia/Controller/Admin/HomeController.php]: Not possible to boolean, because it can be a time value.

**Comment by lunika** [comment]
:+1:

**Pull request merged**